exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/statements/MergeStatement.kt

```diff
@@ -62,20 +75,113 @@ abstract class MergeStatement(val table: Table) : Statement<Int>(
         val and: Op<Boolean>?,
-        /**
-         * deleteWhere is applicable only to Oracle SQL database which has no dedicated "when delete" clause
-         */
-        val deleteWhere: Op<Boolean>? = null
+        /** deleteWhere is applicable only to Oracle SQL database which has no dedicated "when delete" clause */
+        val deleteWhere: Op<Boolean>? = null,
+        /** Postgres clause modifier to override the system value */
+        val overridingSystemValue: Boolean = false,
+        /** Postgres clause modifier to override the user value */
+        val overridingUserValue: Boolean = false,
     )
 
+    enum class ClauseCondition {
+        MATCHED, NOT_MATCHED
+    }
+
     enum class ClauseAction {
-        INSERT, UPDATE, DELETE
+        INSERT, UPDATE, DELETE, DO_NOTHING
+    }
+}
+
+/**
+ * Represents an SQL MERGE statement. It encapsulates the logic to perform conditional updates, insertions,
+ * or deletions.
+ *
+ * Here is only the part specific for the Table as a source implementation.
+ * Look into [MergeStatement] to find the base implementation of that command.
+ *
+ * @property dest The destination [Table] where records will be merged into.
+ * @property source The source [Table] from which records are taken to compare with `dest`.
+ * @property on The join condition [Op<Boolean>] that specifies how to match records in both `source` and `dest`.
+ */
+class MergeTableStatement(
+    dest: Table,
+    private val source: Table,
+    private val on: Op<Boolean>?
+) : MergeStatement(dest) {
+    override fun prepareSQL(transaction: Transaction, prepared: Boolean): String {
+        validateMergeStatement(transaction, clauses)
+        return transaction.db.dialect.functionProvider.merge(table, source, transaction, clauses, on)
+    }
+}
+
+/**
+ * Represents an SQL MERGE statement. It encapsulates the logic to perform conditional updates, insertions,
+ * or deletions.
+ *
+ * Here is only the part specific for the Query as a source implementation.
+ * Look into [MergeStatement] to find the base implementation of that command.
+ *
+ * @property dest The destination [Table] where records will be merged into.
+ * @property selectQuery The source [QueryAlias] from which records are taken to compare with `dest`.
+ * @property on The join condition [Op<Boolean>] that specifies how to match records in both `source` and `dest`.
+ */
+class MergeSelectStatement(
+    dest: Table,
+    private val selectQuery: QueryAlias,
+    val on: Op<Boolean>
+) : MergeStatement(dest) {
+    override fun prepareSQL(transaction: Transaction, prepared: Boolean): String {
+        validateMergeStatement(transaction, clauses)
+        return transaction.db.dialect.functionProvider.mergeSelect(table, selectQuery, transaction, clauses, on, prepared)
+    }
+}
+
+private fun validateMergeStatement(transaction: Transaction, clauses: List<MergeStatement.Clause>) {
+    if (currentDialect !is OracleDialect) {
```

Comment:
Unless there's a specific reason I'm missing, my vote is that this validation function is cut & paste into `FunctionProvider.kt` and called there at the beginning of `merge()` and `mergeSelect()`. That's the current pattern, where all SQL generation logic and dialect-specific logic happens in `FunctionProvider` and its db overrides only.

I think it works because we only have to look in 1 place to see how/why SQL is generated the way it is.
If in the future, for example, another database supports `DO NOTHING` syntax, I'd expect to go and change the logic in that database's function provider (or the default super function provider), not here in the statement file.